utils/env_preferences/src/linux.rs

```diff
@@ -2,70 +2,118 @@
 
-pub mod linux_locales_prefs {
-    use std::{
-        borrow::Cow,
-        collections::{HashMap, HashSet},
-        ffi::CStr,
-        ptr,
-    };
+pub mod linux_prefs {
+    use std::{borrow::Cow, collections::HashMap, ffi::CStr, ptr};
 
     use libc::{setlocale, LC_ALL, LC_TIME};
+    use std::str::FromStr;
 
-    fn fetch_locale_settings() -> HashMap<String, String> {
+    #[derive(Hash, Eq, PartialEq, Debug)]
+    pub enum LocaleCategory {
+        LcCTYPE,
+        LcNUMERIC,
+        LcTIME,
+        LcCOLLATE,
+        LcMONETARY,
+        LcMESSAGES,
+        LcPAPER,
+        LcNAME,
+        LcADDRESS,
+        LcTELEPHONE,
+        LcMEASUREMENT,
+        LcIDENTIFICATION,
+        LcALL,
+        Other(String),
+    }
+
+    impl FromStr for LocaleCategory {
+        type Err = ();
+
+        fn from_str(s: &str) -> Result<Self, Self::Err> {
+            match s {
+                "LC_CTYPE" => Ok(LocaleCategory::LcCTYPE),
+                "LC_NUMERIC" => Ok(LocaleCategory::LcNUMERIC),
+                "LC_TIME" => Ok(LocaleCategory::LcTIME),
+                "LC_COLLATE" => Ok(LocaleCategory::LcCOLLATE),
+                "LC_MONETARY" => Ok(LocaleCategory::LcMONETARY),
+                "LC_MESSAGES" => Ok(LocaleCategory::LcMESSAGES),
+                "LC_PAPER" => Ok(LocaleCategory::LcPAPER),
+                "LC_NAME" => Ok(LocaleCategory::LcNAME),
+                "LC_ADDRESS" => Ok(LocaleCategory::LcADDRESS),
+                "LC_TELEPHONE" => Ok(LocaleCategory::LcTELEPHONE),
+                "LC_MEASUREMENT" => Ok(LocaleCategory::LcMEASUREMENT),
+                "LC_IDENTIFICATION" => Ok(LocaleCategory::LcIDENTIFICATION),
+                "LC_ALL" => Ok(LocaleCategory::LcALL),
+                _ => Ok(LocaleCategory::Other(s.to_string())),
+            }
+        }
+    }
+
+    pub fn get_locales() -> HashMap<LocaleCategory, String> {
         let mut locale_map = HashMap::new();
 
-        // Thread safety is ensured by fallbacking to the default locale of `linux` which is `C`
+        // SAFETY: In case `libc::setlocale` returns a NULL pointer it fallbacks to the default locale "C"
+
         unsafe {
             let locales_ptr = setlocale(LC_ALL, ptr::null());
-            let locales = CStr::from_ptr(locales_ptr);
-
-            if let Ok(locales_str) = locales.to_str() {
-                let locales_slice = locales_str.split(';');
-                for locale in locales_slice {
-                    let mut locale_parts = locale.split('=');
-                    if let (Some(key), Some(value)) = (locale_parts.next(), locale_parts.next()) {
-                        locale_map.insert(key.to_string(), value.to_string());
+            if locales_ptr.is_null() {
+                locale_map.insert(LocaleCategory::LcALL, "C".to_string());
+                return locale_map;
+            }
+
+            // SAFETY: Creating a `CStr` from a non-null pointer and no mutation is being performed.
```

Comment:
`Cstr::from_ptr` has more safety invariants than the pointer being non-null: https://doc.rust-lang.org/std/ffi/struct.CStr.html#method.from_ptr. I at least want to see the null-termination mentioned.